File: frontend/src/utils/mermaidCleaner.js
export function cleanMermaidCode(rawCode) {
  if (!rawCode) return '';

  let code = rawCode
    .replace(/\r\n/g, '\n')
    // Remove wrapping fences if accidentally included
    .replace(/^```mermaid\s*/i, '')
    .replace(/^```\s*/i, '')
    .replace(/```\s*$/i, '')
    // Replace smart/curly quotes with standard quotes
    .replace(/[\u201C\u201D]/g, '"')
    .replace(/[\u2018\u2019]/g, "'")
    // Replace backslashes in paths with forward slashes
    .replace(/\\/g, '/');

  const lines = code.split('\n');
  let subgraphCounter = 1;
  const cleanedLines = [];

  for (let line of lines) {
    let l = line;

    // 1. Remove trailing semicolons
    l = l.replace(/;\s*$/, '');

    // 2. Convert // or # comments to %%
    if (/^\s*\/\//.test(l)) {
      l = l.replace(/^\s*\/\//, '%%');
    } else if (/^\s*#(?!\!)/.test(l)) {
      l = l.replace(/^\s*#/, '%%');
    }

    // 3. Fix subgraph with spaces in ID: e.g. "subgraph Business Logic Services ["Services Layer"]"
    const subgraphWithSpacesMatch = l.match(/^(\s*subgraph\s+)([^"\[\n]+?)(\s*\[.*\])/i);
    if (subgraphWithSpacesMatch) {
      const prefix = subgraphWithSpacesMatch[1];
      const rawId = subgraphWithSpacesMatch[2].trim();
      const bracketPart = subgraphWithSpacesMatch[3];
      if (rawId.includes(' ')) {
        const safeId = rawId.replace(/[^a-zA-Z0-9_]/g, '_');
        l = `${prefix}${safeId}${bracketPart}`;
      }
    }

    // 4. Fix subgraph with quoted title and no ID: e.g. "subgraph "Repository Root: jobfinder/""
    const subgraphQuotedMatch = l.match(/^(\s*subgraph\s+)"([^"]+)"/i);
    if (subgraphQuotedMatch) {
      const prefix = subgraphQuotedMatch[1];
      const title = subgraphQuotedMatch[2];
      const safeId = `sub_${subgraphCounter++}_${title.replace(/[^a-zA-Z0-9_]/g, '_').slice(0, 16)}`;
      l = `${prefix}${safeId} ["${title}"]`;
    }

    // 5. Fix unquoted node labels with special characters like (, ), :, /, & inside [...]
    // Avoid double quoting if already quoted
    l = l.replace(/([a-zA-Z0-9_-]+)\[([^"\]\n]*[\(\):/&][^"\]\n]*)\]/g, '$1["$2"]');

    // 6. Fix unquoted cylinder labels e.g. DB[(MongoDB (Mongoose Models))] -> DB[("MongoDB (Mongoose Models)")]
    l = l.replace(/([a-zA-Z0-9_-]+)\[\(([^"\)\n]+)\)\]/g, '$1[("$2")]');

    // 7. Fix unquoted decision/rhombus labels e.g. Check{Is Score >= Target?} -> Check{"Is Score >= Target?"}
    l = l.replace(/([a-zA-Z0-9_-]+)\{([^"\}\n]*[\(\):/?&><=][^"\}\n]*)\}/g, '$1{"$2"}');

    // 8. Fix unquoted rounded/pill nodes e.g. Start(Process (Init)) -> Start("Process (Init)")
    l = l.replace(/([a-zA-Z0-9_-]+)\(([^"\)\n]*[\(\):/&][^"\)\n]*)\)/g, '$1("$2")');

    cleanedLines.push(l);
  }

  return cleanedLines.join('\n').trim();
}

File: backend/utils/mermaid_cleaner.py
import re

def clean_mermaid_code(raw_code: str) -> str:
    if not raw_code:
        return ""

    code = raw_code.replace("\r\n", "\n")
    code = re.sub(r"^```mermaid\s*", "", code, flags=re.IGNORECASE)
    code = re.sub(r"^```\s*", "", code)
    code = re.sub(r"```\s*$", "", code)
    code = code.replace("\u201c", '"').replace("\u201d", '"')
    code = code.replace("\u2018", "'").replace("\u2019", "'")
    code = code.replace("\\", "/")

    lines = code.split("\n")
    cleaned_lines = []
    subgraph_counter = 1

    for line in lines:
        l = line

        # 1. Remove trailing semicolons
        l = re.sub(r";\s*$", "", l)

        # 2. Convert // or # comments to %%
        if re.match(r"^\s*//", l):
            l = re.sub(r"^\s*//", "%%", l)
        elif re.match(r"^\s*#(?!\!)", l):
            l = re.sub(r"^\s*#", "%%", l)

        # 3. Fix subgraph with spaces in ID
        subgraph_spaces = re.match(r"^(\s*subgraph\s+)([^\"\[\n]+?)(\s*\[.*\])", l, re.IGNORECASE)
        if subgraph_spaces:
            prefix = subgraph_spaces.group(1)
            raw_id = subgraph_spaces.group(2).strip()
            bracket_part = subgraph_spaces.group(3)
            if " " in raw_id:
                safe_id = re.sub(r"[^a-zA-Z0-9_]", "_", raw_id)
                l = f"{prefix}{safe_id}{bracket_part}"

        # 4. Fix subgraph with quoted title and no ID
        subgraph_quoted = re.match(r"^(\s*subgraph\s+)\"([^\"]+)\"", l, re.IGNORECASE)
        if subgraph_quoted:
            prefix = subgraph_quoted.group(1)
            title = subgraph_quoted.group(2)
            safe_id = f"sub_{subgraph_counter}_{re.sub(r'[^a-zA-Z0-9_]', '_', title)[:16]}"
            subgraph_counter += 1
            l = f'{prefix}{safe_id} ["{title}"]'

        # 5. Fix unquoted node labels with special characters
        l = re.sub(r'([a-zA-Z0-9_-]+)\[([^"\]\n]*[\(\):/&][^"\]\n]*)\]', r'\1["\2"]', l)

        # 6. Fix unquoted cylinder labels
        l = re.sub(r'([a-zA-Z0-9_-]+)\[\(([^"\)\n]+)\)\]', r'\1[("\2")]', l)

        # 7. Fix unquoted decision/rhombus labels
        l = re.sub(r'([a-zA-Z0-9_-]+)\{([^"\}\n]*[\(\):/?&><=][^"\}\n]*)\}', r'\1{"\2"}', l)

        # 8. Fix unquoted rounded/pill nodes
        l = re.sub(r'([a-zA-Z0-9_-]+)\(([^"\)\n]*[\(\):/&][^"\)\n]*)\)', r'\1("\2")', l)

        cleaned_lines.append(l)

    return "\n".join(cleaned_lines).strip()

def clean_mermaid_in_markdown(markdown_content: str) -> str:
    if not markdown_content:
        return ""
    
    def replacer(match):
        code = match.group(1)
        cleaned = clean_mermaid_code(code)
        return f"```mermaid\n{cleaned}\n```"

    return re.sub(r"```mermaid\s*([\s\S]*?)```", replacer, markdown_content, flags=re.IGNORECASE)
